gnomad/utils/constraint.py

```diff
@@ -13,6 +13,191 @@
+        agg["singleton_count"] = hl.agg.count_where(singleton_expr)
+
+    for pop in count_downsamplings:
+        logger.info(
+            "Counting variants in downsamplings for population '%s', and adding as 'downsampling_counts_%s' annotation.",
+            pop,
+        )
+        agg[f"downsampling_counts_{pop}"] = downsampling_counts_expr(
+            freq_expr, freq_meta_expr, pop, max_af=max_af
+        )
+        if count_singletons:
+            logger.info(
+                "Counting singleton variants in downsamplings for population '%s', and adding as 'singleton_downsampling_counts_%s' annotation.",
+                pop,
+                pop,
+            )
+            agg[f"singleton_downsampling_counts_{pop}"] = downsampling_counts_expr(
+                freq_expr, freq_meta_expr, pop, singleton=True
+            )
+    # Apply each variant count aggregation in `agg` to get counts for all combinations of `grouping`
+    if use_table_group_by:
+        return ht.group_by(**grouping).partition_hint(partition_hint).aggregate(**agg)
+    else:
+        return ht.aggregate(
+            hl.struct(**{field: hl.agg.group_by(grouping, agg[field]) for field in agg})
+        )
+
+
+def downsampling_counts_expr(
+    freq_expr: hl.expr.ArrayExpression,
+    freq_meta_expr: hl.expr.ArrayExpression,
+    pop: str = "global",
+    variant_quality: str = "adj",
+    singleton: bool = False,
+    max_af: Optional[float] = None,
+) -> hl.expr.ArrayExpression:
+    """
+    Return an aggregation expression to compute an array of counts of all downsamplings found in `freq_expr` where specified criteria is met.
+
+    The meta data (`freq_meta_expr`) should be in a similar format to the `freq_meta` annotation added by  `annotate_freq()`. Each downsampling should have 'group',
+    'pop', and 'downsampling' keys. Included downsamplings are those where 'group' == `variant_quality` and 'pop' == `pop`.
+
+    :param freq_expr: ArrayExpression of Structs with with 'AC' and 'AF' annotations.
+    :param freq_meta_expr: ArrayExpression containing the set of groupings for each element of the `freq_expr` array (e.g., [{'group': 'adj'}, {'group': 'adj', 'pop': 'nfe'}, {'downsampling': '5000', 'group': 'adj', 'pop': 'global'}]).
+    :param pop: Population to use for filtering by the 'pop' key in `freq_meta_expr`. Default is 'global'.
+    :param variant_quality: Variant quality to use for filtering by the 'group' key in `freq_meta_expr`. Default is 'adj'.
+    :param singleton: Whether to filter to only singletons before counting (AC == 1). Default is False.
+    :param max_af: Maximum variant allele frequency to keep. By default no allele frequency cutoff is applied.
+    :return: Aggregation Expression for an array of the variant counts in downsamplings for specified population.
+    """
+    # Get indices of dictionaries in meta dictionaries that only have the "downsampling" key with specified "group" and "pop" values
```

Comment:
Is there a consistent pattern of adding period at the end of each comment? I remember there are some suggestions in the constraint repo about adding period.

Reply:
As discussed, lets move to adding a period at the end of all comments